File: master-test.py
import asyncio
import getpass
import os
import threading

import aiohttp
import dotenv
import websockets

dotenv.load_dotenv()

WS_URL = os.getenv("WS_HOST", "ws://localhost:8000")
HTTP_URL = os.getenv("HTTP_HOST", "http://localhost:8000")


async def login(token: str) -> str:
    async with aiohttp.ClientSession() as session:
        async with session.post(
            f"{HTTP_URL}/api/login",
            json={"token": token},
        ) as resp:
            resp.raise_for_status()
            cookie = resp.cookies.get("session")
            if not cookie:
                raise Exception("No session cookie")
            return cookie.value


async def master(slave_id: str, session_token: str):
    # TODO: Remove token param when go server is complete.
    uri = f"{WS_URL}/ws/master?slave_id={slave_id}&token={session_token}"
    headers = {"Cookie": f"session={session_token}"}
    async with websockets.connect(uri, additional_headers=headers) as websocket:
        print(f"Connected to slave: {slave_id}")

        async def receive_messages():
            try:
                while True:
                    msg = await websocket.recv()
                    if isinstance(msg, str):
                        if msg == "CONNECTED":
                            print("*** Connected to slave ***")
                        elif msg == "DISCONNECTED":
                            print("*** Disconnected from slave ***")
                        elif msg == "HEARTBEAT":
                            print("*** Heartbeat ***")
                        else:
                            print(f"Control: {msg}")
                    else:
                        print(f"Received data: {msg}")
            except Exception:
                pass

        def input_loop():
            while True:
                msg = input("Message: ")
                asyncio.run(websocket.send(msg.encode()))

        threading.Thread(target=input_loop, daemon=True).start()
        await receive_messages()


if __name__ == "__main__":
    auth_token = getpass.getpass("Enter auth token: ")
    session_token = asyncio.run(login(auth_token))
    slave_id = input("Enter slave ID: ").strip()
    asyncio.run(master(slave_id, session_token))

File: tests/master-test.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"net/http"
	"os"
	"strings"

	"github.com/gorilla/websocket"
	"github.com/joho/godotenv"
	"golang.org/x/term"
)

var (
	wsURL   string
	httpURL string
)

func init() {
	godotenv.Load()
	wsURL = os.Getenv("WS_HOST")
	if wsURL == "" {
		wsURL = "ws://localhost:8080"
	}
	httpURL = os.Getenv("HTTP_HOST")
	if httpURL == "" {
		httpURL = "http://localhost:8080"
	}
}

func login(token string) (string, error) {
	resp, err := http.Post(httpURL+"/api/login", "application/json", strings.NewReader(fmt.Sprintf(`{"token":"%s"}`, token)))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("login failed with status: %d", resp.StatusCode)
	}

	cookies := resp.Cookies()
	for _, c := range cookies {
		if c.Name == "session" {
			return c.Value, nil
		}
	}
	return "", fmt.Errorf("no session cookie found")
}

func connectMaster(slaveID, sessionToken string) {
	header := http.Header{}
	header.Add("Cookie", fmt.Sprintf("session=%s", sessionToken))

	urlStr := fmt.Sprintf("%s/ws/master?slave_id=%s", wsURL, slaveID)
	conn, _, err := websocket.DefaultDialer.Dial(urlStr, header)
	if err != nil {
		log.Fatalf("Failed to connect: %v", err)
	}
	defer conn.Close()

	fmt.Printf("Connected to slave: %s\n", slaveID)

	go func() {
		for {
			_, msg, err := conn.ReadMessage()
			if err != nil {
				return
			}
			fmt.Printf("Received: %s\n", msg)
		}
	}()

	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Print("Message: ")
		if !scanner.Scan() {
			return
		}
		input := scanner.Text()
		if err := conn.WriteMessage(websocket.TextMessage, []byte(input)); err != nil {
			return
		}
	}
}

func main() {
	fmt.Print("Enter auth token: ")
	tokenBytes, _ := term.ReadPassword(int(os.Stdin.Fd()))
	token := string(tokenBytes)
	fmt.Println()

	sessionToken, err := login(token)
	if err != nil {
		log.Fatalf("Login failed: %v", err)
		os.Exit(1)
	}
	fmt.Println("Login successful")

	var slaveID string
	fmt.Print("Enter slave ID: ")
	fmt.Scanln(&slaveID)

	connectMaster(slaveID, sessionToken)
}
